pages/tutorials/3.0/window/events.md

```diff
@@ -7,259 +7,329 @@
+SFML uses `std::variant` under the hood to implement a type-safe API for events.
+There are two main ways to use this API.
+Check out the EventHandling example program to see these methods in practice.
 
-[`sf::Event`](https://www.sfml-dev.org/documentation/3.0.0/classsf_1_1Event.php "sf::Event documentation") instances are filled by the `pollEvent` (or `waitEvent`) function of the [`sf::Window`](https://www.sfml-dev.org/documentation/3.0.0/classsf_1_1Window.php "sf::Window documentation") class. Only these two functions can produce valid events, any attempt to use an [`sf::Event`](https://www.sfml-dev.org/documentation/3.0.0/classsf_1_1Event.php "sf::Event documentation") which was not returned by successful call to `pollEvent` (or `waitEvent`) will result in the same undefined behavior that was mentioned above.
+### `sf::Event::getIf<T>`
 
-To be clear, here is what a typical event loop looks like:
+The first option is based around `sf::Event::getIf<T>` and `sf::Event::is<T>`.
+`getIf<T>` works by providing a template parameter which must be an event subtype.
+Event subtypes are types like `sf::Event::Resized` or `sf::Event::MouseMoved`.
+If the template argument matches the active event subtype then a pointer to the data is returned.
+If that template argument is not the active event subtype then `nullptr` is returned.
+`sf::Event::is<T>` more simply returns `true` if `T` matches the active event subtype.
+`is<T>` is often used for subtypes like `sf::Event::Closed` which contain no data.
+Here's what that looks like:
 
 ```cpp
-sf::Event event;
-
-// while there are pending events...
-while (window.pollEvent(event))
+while (window.isOpen())
 {
-    // check the type of the event...
-    switch (event.type)
+    while (const std::optional event = window.pollEvent())
     {
-        // window closed
-        case sf::Event::Closed:
+        if (event->is<sf::Event::Closed>())
+        {
             window.close();
-            break;
-
-        // key pressed
-        case sf::Event::KeyPressed:
-            ...
-            break;
-
-        // we don't process other types of events
-        default:
-            break;
+        }
+        else if (const auto* keyPressed = event->getIf<sf::Event::KeyPressed>())
+        {
+            if (keyPressed->scancode == sf::Keyboard::Scancode::Escape)
+                window.close();
+        }
     }
+
+    // Remainder of main loop
 }
 ```
 
-Read the above paragraph once again and make sure that you fully understand it, the [`sf::Event`](https://www.sfml-dev.org/documentation/3.0.0/classsf_1_1Event.php "sf::Event documentation") union is the cause of many problems for inexperienced programmers.
+Note how the API for getting events has changed slightly.
+`sf::WindowBase::pollEvent` and `sf::WindowBase::waitEvent` now return a `std::optional<sf::Event>`.
```

Comment:
I think this should be at the topic, similar to how it was before. It makes sense to me, that we first show how to get events and then explain what you can do with them.